ce/api/streamflow/watershed_streams.py

```diff
@@ -0,0 +1,171 @@
+
+    This function exists to make these computations more easily testable.
+    (Specifically, data *files* are not required, only the relevant contents
+    of those files passed as `VicDataGrid` objects. `VicDataGrid`s are much easier to
+    construct for tests.)
+
+    :param station_lonlat: (tuple) Location of drainage point, (lon, lat)
+    :param flow_direction: (VicDataGrid) Flow direction grid
+    :return: (dict) representation for JSON response object; see watershed() for details
+    """
+
+    # Compute lonlats of watershed whose mouth is at `station`
+    # TODO: Refactor to accept a VicDataGrid?
+    direction_matrix = VIC_direction_matrix(
+        flow_direction.lat_step, flow_direction.lon_step
+    )
+
+    with Timer() as watershed_time:
+        watershed_xys = build_watershed_streams(
+            flow_direction.lonlat_to_xy(station_lonlat),
+            flow_direction.values,
+            direction_matrix,
+            debug=True,
+        )
+
+    # `watershed_lonlats` must be an ordered collection (not sets) because
+    # a multi line string is an array (python list) of linestrings
+    watershed_lonlats = [[]]
+    counter = 0
+
+    for frozen in watershed_xys:
+        for xy in frozen:
+            watershed_lonlats[counter].append(flow_direction.xy_to_lonlat(xy))
+        counter += 1
+        watershed_lonlats.append([])
+
+    if watershed_lonlats[-1] == []:
+        watershed_lonlats.remove([])
+
+    lines = MultiLineString(watershed_lonlats)
+
+    return {
+        "streams": geojson_feature(lines,),
+        "debug/test": {"watershed_streams": {"time": watershed_time.elapsed,}},
+    }
+
+
+def build_watershed_streams(target, routing, direction_map, debug=False):
+    """
+    Return set of all paths of cells that drain into `target`.
+
+    :param target: An xy index representing the cell of interest.
+    :param routing: A numpy array representing water flow using VIC direction
+        codes (0 - 9). These codes may be integers or floats.
+    :param direction_map: Maps VIC direction codes into offsets in cell indices.
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+        TODO: Compute this internally?
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: Set of frozensets of cells (streams) that drain into `target`.
+
+    Notes:
+
+    - In this function, a cell is represented by an (x, y) index pair.
+
+    - Routing graphs can and in practice do contain cycles. Variable `visited`
+    is used to determine whether a cell has already been visited during the
+    traversal of the routing graph, i.e., whether we are cycling, and if so
+    not to repeat that subgraph.
+    """
+    visited = set()
+    i = 0
+    connection = [[]]
+
+    def upstream(cell):
```

Comment:
I'm assuming that we are continuing with the "long streams" approach as opposed to the "length-2 line segments" approach discussed in the issue.

I'm having a difficult time deciding if this function does what (I think) it is supposed to do. It's not at all obvious to me whether the recursive calls to `upstream` leave `i` and `connection` in correct states in all cases. Can you explain how that works?

Reply:
When a cell has an eligible neighbor, a recursive call is made on the eligible neighbor. This will cause all of the cells on one path to be put into a list. When ```connection.append([])``` happens, this means that there are no longer any eligible upstream cells (since we will have returned from the consistent calls of ```upstream(neighbour)```). We then start a new path from the furthest point upstream that has a neighbor. This cell is added to the path, and we continue the process, until it returns it's way back down to the initial cell. After the ```upstream``` is done running, connection contains a list of lists, the inner lists being the path's of cells. I made a mistake overloading ```i```. I have updated it so that the variable used in ```upstream``` is called counter, as this variable counts the number of paths in ```connection``` (so it knows where to put the next cell that is upstream). The ```i``` that was used in the return statement has been renamed to ```stream```, as this is a more fitting name.

Reply:
that is correct

Reply:
We apparently have an algorithm that works (tests pass :tada:), and we have an explanation of how it works.

That it works is good. Now we have to ask how readable and maintainable it is, because it is part of a living codebase. My thoughts on this, in descending order of importance:

1. It has what's often called a "code smell", which in this case is the creation of unused objects (`connection`s) which have to be cleaned up at the end. Creating things that have no use, even transiently, suggest a tighter algorithm can be developed.
2. Add comments that explain how the algorithm works. It's not instantly obvious. I suggest a sort of global explanation akin to the ones you and I formulated, plus the occasional comment on a key line in the body of the code.
3. Some algorithms need explanations because they are inherently tricky, or rather because the problem they solve forces some trickiness. This may be one such. But at this point, we should wonder if there is an algorithm whose operation is clearer and which needs less explanation. The benefit being to later maintainers who need to modify what it does. Which leads into the next observation: 
4. This is an example of a recursive algorithm with global (nonlocal) variables that it _modifies_ -- they are the result it generates. 
   - That kind of algorithm is notoriously tricky to write and understand: evidenced by the fact that I had to spend a few minutes thinking about how it works and if it works in all cases and to confirm that understanding with you. 
   - Recursive algorithms are more "natural" and comprehensible when they don't use global variables and instead restrict themselves to values they receive in arguments. These are called purely functional (no side effects; no global variables). They construct their results from only their arguments and the return values of recursive calls. You only have to establish that the algorithm works on the local information not on hard to establish guarantees that the global variables are always in a suitable state at each recursive call. 
   - There is (by first principles) an equivalent purely functional algorithm. Whether such an algorithm would actually be simpler to write and understand is an open question, but it's worth giving it some thought. An exercise for the student, so to speak.
5. Finally, a bit around naming: Leaving aside the question of whether we go to a purely functional algorithm, we have variables named `connection` and `counter`. But when you explain the algorithm, you use the terms "streams" and "number of streams". It's always better for the explanation and the code to be congruent, i.e., to use the same terminology.

What I suggest at this point is the following:
- Set a deadline of 1 week for any renovations to this algorithm plus the other few outstanding comments/changes. If there are still questions, problems, or tests not passing, then we go back to the current algorithm.
- Fix items 1 and 2 above.
- Think about items 3 and 4, and at least outline your thinking to me. This is intended to be a learning experience -- both algorithmically and about what it means to participate in a long-lived codebase. If we agree it's worthwhile, we can update the code accordingly.
- Do 5 as appropriate to the algorithm finally arrived at.